[incubator/mysqlha] Add metrics container

holmsten opened the pull request:
Add 'prom/mysqld_exporter' sidecar container to expose metrics in a format that Prometheus can scrape. Enable exporter by setting `metrics.enable=true` when deploying the chart. Port 9104 is exposed when metrics are enabled.

holmsten committed:
mysqlha: Add metrics container

Add 'prom/mysqld_exporter' sidecar container to expose metrics in a
format that Prometheus can scrape. Enable exporter by setting
`metrics.enable=true` when deploying the chart. Port 9104 is exposed
when metrics are enabled.

Bump mysqlha version to 0.1.2

holmsten committed:
mysqlha: Add metrics documentation in README.md

Add parameter documentation for metrics sidecar configuration options

holmsten commented:
/assign @sameersbn

sameersbn commented on incubator/mysqlha/values.yaml:
```diff
@@ -39,6 +39,16 @@ persistence:
+  imageTag: v0.10.0
+  imagePullPolicy: IfNotPresent
+  resources: {}
+  annotations:
```
Looks like you missed updating the templates to use the annotations.

holmsten replied:
Good catch, thanks! Updated template accordingly.

sameersbn commented on incubator/mysqlha/templates/statefulset.yaml:
```diff
@@ -215,6 +215,35 @@ spec:
           requests:
             cpu: 100m
             memory: 100Mi
+      {{- if .Values.metrics.enabled }}
```
since the metrics endpoint will report the global status of the cluster, I am not sure if we should add a exporter for each pod as opposed to having a single exporter that talks to the master node to gather the metrics.

holmsten replied:
There are some slave-specific metrics that you might want to analyse so only running on master feels counterproductive.

sameersbn replied:
I am not very knowledgeable of the metrics the mysql exporter exposes, but looking at the documentation at https://github.com/prometheus/mysqld_exporter it appears a config named `collect.slave_status` (Collect from SHOW SLAVE STATUS) is enabled by default. That's the reason I was wondering if it really made sense to have as many mysqld_exporters as there are mysqld pods. 

Additionally in svc.yaml the metrics port is exposed only for the `write` endpoint (service with the name `{{ template "fullname" . }}`) which means that the exporters associated with the readonly pods will not be reached anyway.

@jpoon in svc.yaml i see two services exist `{{ template "fullname" . }}` and `{{ template "fullname" . }}-readonly`. i was wondering how the pods are filtered here.

timstoop replied:
As slaves are used for read request, you want to gather metrics from them separately. Metrics exposed are things you'd use if you run into performance problems, for instance. For that, they need to be available per instance.

timstoop replied:
To add to this, `collect.slave_status` just adds the metrics it can collect from SHOW SLAVE STATUS, but that's mostly just how much lag a slave has with the master.

sameersbn reviewed (changes requested):
/unassign 
/assign @jpoon

sameersbn commented:
@jpoon could you please take a look here. thanks!

jpoon commented:
Should we create a new user for the exporter? Aside from that LGTM.

holmsten committed:
Add metrics annotations

Add metrics annotations to service if specified as a configuration value

holmsten commented:
I reused much code from mariadb chart where root user is used for exporter. I can agree that it's better to create a new user. If you feel that this is a requirement I can make the changes required.

carrodher commented:
It seems that in the meantime a new version has appeared, can you increase the version again?

holmsten committed:
Bump chart version to 0.1.3

holmsten committed:
Merge branch 'master' into mysqlha-exporter

jpoon commented:
Yeah, it's probably best to create a new user. There should already be a script in there to do that, you can copy the logic from `create-replication-user.sql`

holmsten commented:
Copying the logic is fine but mysqld_exporter docker image doesn't ship with mysql client, and I don't want to maintain a docker image. Using official image feels like a better solution in my opinion. User creation can be performed in the xtrabackup container, but that doesn't feel like a good enough solution. Thoughts?

timstoop commented on incubator/mysqlha/Chart.yaml:
```diff
@@ -1,5 +1,5 @@
 name: mysqlha
-version: 0.1.2
+version: 0.1.3
```
As you're adding a new feature, please bump the minor version.

timstoop commented:
I don't see a problem with creating the user from the xtrabackup container if it helps keep the number of unique containers down. Just my opinion, though.

holmsten committed:
mysqlha: Add creation of metrics mysql user

* Add creation of metrics mysql user

mysqld_exporter container should run using a user with only a selected
set of grants required for operation, instead of root user.

* Add create-metrics-user.sql to configmap

* Add metrics mysql user credentials to secrets

* Configure metrics mysql user in init-mysql and xtrabackup container

If metrics collection is enabled the user creation sql query is copied
from configmap to scripts volume when running the init-mysql
initContainer. Inside xtrabackup container and once the mysql container
is running, the metrics mysql user is created with credentials fetched
from secrets.

* Add parameter documentation for metrics mysql user configuration options

* Increase Chart version to 0.2.0

timstoop commented:
/ok-to-test
/lgtm

holmsten committed:
Merge branch 'master' into mysqlha-exporter

holmsten committed:
Merge branch 'master' into mysqlha-exporter

unguiculus commented:
cc @gianrubio

gianrubio commented on incubator/mysqlha/templates/statefulset.yaml:
```diff
@@ -149,6 +155,13 @@ spec:
             secretKeyRef:
               name: {{ template "fullname" . }}
               key: mysql-replication-password
+        - name: MYSQL_METRICS_USER
```
Please add ``` {{- if .Values.metrics.enabled }} ``` here, ending on line 164 otherwise the deployment will fail because the secret doesn't exist

gianrubio reviewed (changes requested):
Just a minor change, otherwise the changes are ok to me

holmsten committed:
Merge remote-tracking branch 'upstream/master' into mysqlha-exporter

holmsten committed:
[incubator/mysqlha] Add check for mysqlMetricsUser

* Add check for mysqlMetricsUser

Only add metrics environment variables to xtrabackup container if
mysqlMetricsUser is defined

* Bump chart version to 0.3.0

holmsten committed:
[incubator/mysqlha] Bump Chart version to 0.3.0

holmsten committed:
[incubator/mysqlha] Remove unexpected `{{end}}`

During merge an extra `{{end}}` was mistakenly added causing a parse
error, this commit remove that extra line

holmsten committed:
Merge branch 'master' of https://github.com/kubernetes/charts into mysqlha-exporter
